Swagger changes for Microsoft.Migrate namespace

shsingla opened the pull request:
### Latest improvements:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i><br>
### Contribution checklist:
- [ ] I have reviewed the [documentation](https://github.com/Azure/azure-rest-api-specs#basics) for the workflow.
- [ ] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.
- [ ] The [OpenAPI Hub](https://aka.ms/openapiportal) was used for checking validation status and next steps.
### ARM API Review Checklist
- [ ] Service team MUST add the "WaitForARMFeedback" label if the management plane API changes fall into one of the below categories. 
- adding/removing APIs.
- adding/removing properties.
- adding/removing API-version. 
- adding a new service in Azure.

Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs.
- [ ] If you are blocked on ARM review and want to get the PR merged urgently, please get the ARM oncall for reviews (RP Manifest Approvers team under Azure Resource Manager service) from IcM and reach out to them. 
Please follow the link to find more details on [API review process](https://armwiki.azurewebsites.net/rp_onboarding/ResourceProviderOnboardingAPIRevieworkflow.html).

shsingla committed:
Create migrate.json

shsingla committed:
Adding some examples.

Adding some examples.

MyronFanQiu commented:
@shsingla Hello. Please solve the CI errors.

shsingla committed:
Fixing CI errors.

Fixing CI errors.

shsingla commented:
Hello @MyronFanQiu I have fixed some of the CI errors though some are left as I need to attach the examples files with all status codes. Can you please start to take a look into the PR by the time I will also attach the example files.

shsingla committed:
Fixing spelling issues.

Fixing spelling issues.

shsingla commented on specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:
```diff
@@ -0,0 +1,3748 @@
+          "description": "Monthly storage cost estimate for the disks that are attached to this machine as a group, for a 31-day month.",
+          "readOnly": true
+        },
+        "montlyPremiumStorageCost": {
```
Done.

shsingla commented on specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:
```diff
@@ -0,0 +1,3748 @@
+          "enum": [
+            "Unknown",
+            "NotApplicable",
+            "InternalErrorOccured"
```
Done.

MyronFanQiu commented on specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:
```diff
@@ -0,0 +1,3748 @@
+        },
+        "assessmentSolutionId": {
+          "type": "string",
+          "description": "Assessment solution ARM id tracked by Microsoft.Migrate/migrateProjects."
```
Is this readonly? Could user set this field?

shsingla replied:
No this is not read only. User can set this field.

MyronFanQiu commented on specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:
```diff
@@ -0,0 +1,3748 @@
+          "x-nullable": true,
+          "readOnly": true
+        },
+        "provisioningState": {
```
Could user set this field?

shsingla replied:
No user cannot set this. Making it read only.

MyronFanQiu commented on specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:
```diff
@@ -0,0 +1,3748 @@
+      "type": "object",
+      "properties": {
+        "authority": {
+          "type": "string"
```
Could we add description for these propeties?

shsingla replied:
Done.

MyronFanQiu commented on specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:
```diff
@@ -0,0 +1,3748 @@
+      }
+    },
+    "AssessmentOptionsResultList": {
+      "type": "object",
```
Write 404 in the responses may casue some errors for SDK generation process. I may need to check the details again and let you know the conclusion.

MyronFanQiu replied:
@shsingla Hello. Usually, we will just use default as the error response. Is there any specific reason you need to use 404?

shsingla replied:
Done moved to default.

shsingla committed:
Taking Comments.

pilor commented:
>           "$ref": "#/definitions/ProjectProperties"

"x-ms-client-flatten" will improve the experience for users specifying these properties (the old api-version used it)

---
Refers to: specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:164 in f533bc5. [](commit_id = f533bc571046c13e488c2e4a8df0391a1ed6d3ef, deletion_comment = False)

pilor commented:
>           "description": "URL of service map workspace created by customer."

The description says "URL" but the property is named "WorkspaceId".  Is this a resource ID or a URL?

---
Refers to: specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:207 in f533bc5. [](commit_id = f533bc571046c13e488c2e4a8df0391a1ed6d3ef, deletion_comment = False)

pilor commented:
>     "/subscriptions/{subscriptionId}/resourcegroups/{resourceGroupName}/providers/Microsoft.Migrate/assessmentProjects": {

is the existing "projects" resourceType changing names?  How are the existing resources being migrated?

---
Refers to: specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:2065 in f533bc5. [](commit_id = f533bc571046c13e488c2e4a8df0391a1ed6d3ef, deletion_comment = False)

pilor commented:
>         "operationId": "Projects_List",

You need to support listing all assessmentProjects in the subscription as well ([https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/resource-api-reference.md#get-resource](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/resource-api-reference.md#get-resource))

---
Refers to: specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:2069 in f533bc5. [](commit_id = f533bc571046c13e488c2e4a8df0391a1ed6d3ef, deletion_comment = False)

pilor commented:
>       "post": {

why is modification of machines in a group handled via a POST action?  How does a user see what machines are in the group?  Could the group resource just have an array of machines in its body?

---
Refers to: specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:2758 in f533bc5. [](commit_id = f533bc571046c13e488c2e4a8df0391a1ed6d3ef, deletion_comment = False)

pilor commented:
>     "GroupProperties": {

Its hard to review a lot of these changes without examples being present.  Please include examples for all APIs in this spec.

---
Refers to: specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:290 in f533bc5. [](commit_id = f533bc571046c13e488c2e4a8df0391a1ed6d3ef, deletion_comment = False)

shsingla commented:
> > ```
> > "/subscriptions/{subscriptionId}/resourcegroups/{resourceGroupName}/providers/Microsoft.Migrate/assessmentProjects": {
> > ```
> 
> is the existing "projects" resourceType changing names? How are the existing resources being migrated?
> 
> Refers to: specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:2065 in [f533bc5](https://github.com/Azure/azure-rest-api-specs/commit/f533bc571046c13e488c2e4a8df0391a1ed6d3ef). [](commit_id = [f533bc5](https://github.com/Azure/azure-rest-api-specs/commit/f533bc571046c13e488c2e4a8df0391a1ed6d3ef), deletion_comme

> > ```
> >     "operationId": "Projects_List",
> > ```
> 
> You need to support listing all assessmentProjects in the subscription as well (https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/resource-api-reference.md#get-resource)
> 
> Refers to: specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:2069 in [f533bc5](https://github.com/Azure/azure-rest-api-specs/commit/f533bc571046c13e488c2e4a8df0391a1ed6d3ef). [](commit_id = [f533bc5](https://github.com/Azure/azure-rest-api-specs/commit/f533bc571046c13e488c2e4a8df0391a1ed6d3ef), deletion_comment = False)

Done. It is supported was not added to swagger, adding it now.

shsingla commented:
> > ```
> >     "operationId": "Projects_List",
> > ```
> 
> You need to support listing all assessmentProjects in the subscription as well (https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/resource-api-reference.md#get-resource)
> 
> Refers to: specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:2069 in [f533bc5](https://github.com/Azure/azure-rest-api-specs/commit/f533bc571046c13e488c2e4a8df0391a1ed6d3ef). [](commit_id = [f533bc5](https://github.com/Azure/azure-rest-api-specs/commit/f533bc571046c13e488c2e4a8df0391a1ed6d3ef), deletion_comment = False)

Done. It is supported, adding to swagger.

shsingla commented:
> > ```
> >       "description": "URL of service map workspace created by customer."
> > ```
> 
> The description says "URL" but the property is named "WorkspaceId". Is this a resource ID or a URL?
> 
> Refers to: specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:207 in [f533bc5](https://github.com/Azure/azure-rest-api-specs/commit/f533bc571046c13e488c2e4a8df0391a1ed6d3ef). [](commit_id = [f533bc5](https://github.com/Azure/azure-rest-api-specs/commit/f533bc571046c13e488c2e4a8df0391a1ed6d3ef), deletion_comment = False)

It should be the resource id. Updating the description.

shsingla commented:
> > > ```
> > >       "description": "URL of service map workspace created by customer."
> > > ```
> > 
> > 
> > The description says "URL" but the property is named "WorkspaceId". Is this a resource ID or a URL?
> > Refers to: specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:207 in [f533bc5](https://github.com/Azure/azure-rest-api-specs/commit/f533bc571046c13e488c2e4a8df0391a1ed6d3ef). [](commit_id = [f533bc5](https://github.com/Azure/azure-rest-api-specs/commit/f533bc571046c13e488c2e4a8df0391a1ed6d3ef), deletion_comment = False)
> 
> It should be the resource id. Updating the description.



> > > ```
> > > "/subscriptions/{subscriptionId}/resourcegroups/{resourceGroupName}/providers/Microsoft.Migrate/assessmentProjects": {
> > > ```
> > 
> > 
> > is the existing "projects" resourceType changing names? How are the existing resources being migrated?
> > Refers to: specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:2065 in [f533bc5](https://github.com/Azure/azure-rest-api-specs/commit/f533bc571046c13e488c2e4a8df0391a1ed6d3ef). [](commit_id = [f533bc5](https://github.com/Azure/azure-rest-api-specs/commit/f533bc571046c13e488c2e4a8df0391a1ed6d3ef), deletion_comme
> 
> > > ```
> > >     "operationId": "Projects_List",
> > > ```
> > 
> > 
> > You need to support listing all assessmentProjects in the subscription as well (https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/resource-api-reference.md#get-resource)
> > Refers to: specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:2069 in [f533bc5](https://github.com/Azure/azure-rest-api-specs/commit/f533bc571046c13e488c2e4a8df0391a1ed6d3ef). [](commit_id = [f533bc5](https://github.com/Azure/azure-rest-api-specs/commit/f533bc571046c13e488c2e4a8df0391a1ed6d3ef), deletion_comment = False)



> > ```
> > "GroupProperties": {
> > ```
> 
> Its hard to review a lot of these changes without examples being present. Please include examples for all APIs in this spec.
> 
> Refers to: specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:290 in [f533bc5](https://github.com/Azure/azure-rest-api-specs/commit/f533bc571046c13e488c2e4a8df0391a1ed6d3ef). [](commit_id = [f533bc5](https://github.com/Azure/azure-rest-api-specs/commit/f533bc571046c13e488c2e4a8df0391a1ed6d3ef), deletion_comment = False)

Adding some more examples. Will be adding the rest ASAP.

shsingla committed:
Taking CR comments.

Taking CR comments.

1. Adding the get assessmentProjects at the subscription level.
2. Removing error response code and adding default instead.
3. Updating some descriptions.

shsingla commented:
> > ```
> > "/subscriptions/{subscriptionId}/resourcegroups/{resourceGroupName}/providers/Microsoft.Migrate/assessmentProjects": {
> > ```
> 
> is the existing "projects" resourceType changing names? How are the existing resources being migrated?
> 
> Refers to: specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:2065 in [f533bc5](https://github.com/Azure/azure-rest-api-specs/commit/f533bc571046c13e488c2e4a8df0391a1ed6d3ef). [](commit_id = [f533bc5](https://github.com/Azure/azure-rest-api-specs/commit/f533bc571046c13e488c2e4a8df0391a1ed6d3ef), deletion_comment = False)



> > ```
> >   "post": {
> > ```
> 
> why is modification of machines in a group handled via a POST action? How does a user see what machines are in the group? Could the group resource just have an array of machines in its body?
> 
> Refers to: specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:2758 in [f533bc5](https://github.com/Azure/azure-rest-api-specs/commit/f533bc571046c13e488c2e4a8df0391a1ed6d3ef). [](commit_id = [f533bc5](https://github.com/Azure/azure-rest-api-specs/commit/f533bc571046c13e488c2e4a8df0391a1ed6d3ef), deletion_comment = False)

This was done to handle the scale scenario. We currently are supporting a scale where a group can contain upto 35000 machines. So while creating a group it was not possible to send all the 35000 machines in one go as it exceeds the ARM limit so we went with this approach. Sharing an earlier mail thread over email where this has been discussed.

shsingla commented:
> > ```
> > "/subscriptions/{subscriptionId}/resourcegroups/{resourceGroupName}/providers/Microsoft.Migrate/assessmentProjects": {
> > ```
> 
> is the existing "projects" resourceType changing names? How are the existing resources being migrated?
> 
> Refers to: specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:2065 in [f533bc5](https://github.com/Azure/azure-rest-api-specs/commit/f533bc571046c13e488c2e4a8df0391a1ed6d3ef). [](commit_id = [f533bc5](https://github.com/Azure/azure-rest-api-specs/commit/f533bc571046c13e488c2e4a8df0391a1ed6d3ef), deletion_comment = False)

The existing "projects" resourceType is not changing names. The "assessmentProjects" is a new resource of our service for a V2 experience which is more integrated. The "projects" resourceType still exists and will be deprecated soon.

shsingla committed:
Adding the remaining examples.

Adding the remaining examples.

pilor commented on specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:
```diff
@@ -0,0 +1,3778 @@
+        },
+        "name": {
+          "type": "string",
+          "read
```
consider making this "listDownloadUrl".  I know it sounds odd but it lets people consume it from a template deployment.  Existing example:  https://docs.microsoft.com/en-us/rest/api/containerregistry/runs/getlogsasurl

shsingla replied:
Actually, we will always be having a single URL so we are not thinking of making it as a list. Moreover changing this now will lead to a lot of breaking changes for us. So it would be fine for now to keep it as is.

pilor commented on specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:
```diff
@@ -0,0 +1,3778 @@
+            "modelAsString": true
+          }
+        },
+        "suitabilityExplanation": {
```
The difference between suitabilityExplanation and suitabilityDetail is not clear from the property names or descriptions

pilor commented on specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:
```diff
@@ -0,0 +1,3778 @@
+    "CollectorProperties": {
+      "type": "object",
+      "properties": {
+        "discoverySiteId": {
```
Everything in here needs descriptions

pilor commented on specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:
```diff
@@ -0,0 +1,3778 @@
+    "CollectorAgentProperties": {
+      "type": "object",
+      "properties": {
+        "id": {
```
Having an "id" property inside of properties gets confusing since the resource itself has a top-level "id".  This should be named something else if possible

pilor commented on specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:
```diff
@@ -0,0 +1,3778 @@
+        }
+      }
+    },
+    "CollectorProperties": {
```
hyperV and vmWare collectors seem to have the same properties.  If collectors are going to generally look the same you can have one "collectors" resourceType and differentiate them using the "kind" property.  See https://github.com/Azure/autorest/blob/master/docs/extensions/readme.md#x-ms-discriminator-value (and search x-ms-discrimanator for examples in the repo)

pilor commented on specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:
```diff
@@ -0,0 +1,3778 @@
+          "description": "Intended audience for the service principal.",
+          "type": "string"
+        },
+        "objectId": {
```
Is there a reason that MSIs can't be used for this scenario?  The identity would be represented in Azure by a resource.  https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/overview

pilor replied:
Can you expand on that?  MSIs are the way that identities used by a resource should be handled in Azure.  Can you explain how specifying an explicit objectId changes the need to grant that identity permissions to whatever this is acting on?

shsingla replied:
The earlier comment describes why we didn't use MSIs for our S2S between our services. Attaching a mail for that to the mail thread that is going on. Misread your comment earlier. Now for this code, we are storing all these spn details for the agent. These details will be stored in our DB. Now when an agent gets a token from AAD and makes a call to our service we will verify the token against these stored details. We don't know a way where MSIs can work with an agent in our scenario. Moreover, the same pattern is being used in our other service i.e discovery service (which has already been approved in arm review) so we wanted to follow it as a standard paradigm across our services.

shsingla commented on specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:
```diff
@@ -0,0 +1,3778 @@
+        },
+        "name": {
+          "type": "string",
+          "read
```
This api is actually not paginated. Removing x-ms-pageable.

pilor replied:
That doesn't change the requirements on the format.  Since this is a collection GET it should wrap the result in a "value": [] array and if this is a singleton response represent the single assessmentOption as a resource with name "default".  Your other option is to model this as a POST action called listAssessmentOptions.  The POST action would let you keep the current response but you would need to map the authorization action from /action to /read in your manifest so users could still call it if the resource was locked

shsingla replied:
Actually this is not a collection get this is a single get. I think the name is causing the confusion, renamed it appropriately.

shsingla commented on specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:
```diff
@@ -0,0 +1,3778 @@
+        "createdTimestamp": {
+          "type": "string",
+          "format": "date-time",
+          "description": "Time when this project was created. Date-Time represented in ISO-8601 format.",
```
Done.

shsingla committed:
Taking commenst

Taking comments.

pilor commented on specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:
```diff
@@ -0,0 +1,3778 @@
+        },
+        "name": {
+          "type": "string",
+          "read
```
assessedMachines are under assessments which are under groups... Is the design here that that only way I can see what machines are in my group is by running an assessment, waiting for it to complete, then looking at assessedMachines?  Why not expose the list of machines in a way that is easily viewable?

shsingla replied:
We also have a get machines API which can show the machines under the group by passing group as a filter param.

pilor commented on specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:
```diff
@@ -0,0 +1,3777 @@
+          "type": "string",
+          "description": "The ARM id of service map workspace created by customer."
+        },
+        "customerWorkspaceLocation": {
```
why does the user need to provide this?  What happens if they provide an inaccurate value?  If the RP needs the location it should get it from the resource.

shsingla replied:
The customer workspace actually is the Azure service map workspace. The user needs to provide the location so that we know to which service (Azure service bus) endpoint we need to send the call to. If this is provided as inaccurate the call to the service endpoint will fail. The RP is not owning this resource so we can't get its location.

pilor reviewed (changes requested):
There were some existing comments that weren't resolved above as well

shsingla committed:
Taking comments.

Taking comments.

majastrz commented:
>         "properties": {

This used to have x-ms-flatten in the previous API version. This will result in SDK-breaking changes. Was it intentional?

---
Refers to: specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:162 in 8acc868. [](commit_id = 8acc8688c38f5db44352e7d4336e8e89927f0c87, deletion_comment = False)

majastrz commented:
>             "Inactive"

Will the service continue returning the old enum values with the old API version?

---
Refers to: specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:197 in 8acc868. [](commit_id = 8acc8688c38f5db44352e7d4336e8e89927f0c87, deletion_comment = False)

majastrz commented:
>         "properties": {

x-ms-flatten was removed from this definition as well.

---
Refers to: specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:284 in 8acc868. [](commit_id = 8acc8688c38f5db44352e7d4336e8e89927f0c87, deletion_comment = False)

majastrz commented:
>         "areAssessmentsRunning": {

It's not required, but have you considered making it an enum. If it was named `assessmentStatus`, you could easily add more values in the future.

---
Refers to: specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:326 in 8acc868. [](commit_id = 8acc8688c38f5db44352e7d4336e8e89927f0c87, deletion_comment = False)

majastrz commented:
>     "/subscriptions/{subscriptionId}/resourcegroups/{resourceGroupName}/providers/Microsoft.Migrate/assessmentProjects": {

Yes, I'd like to know as well. `projects` is a tracked resource also.

---
In reply to: [528015267](https://github.com/Azure/azure-rest-api-specs/pull/7063#issuecomment-528015267) [](ancestors = 528015267)

---
Refers to: specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:2065 in f533bc5. [](commit_id = f533bc571046c13e488c2e4a8df0391a1ed6d3ef, deletion_comment = False)

majastrz commented:
>           "nextLinkName": null

Even if the service doesn't return `nextLink`, you should declare it in the swagger. This way adding paging later is much easier.

---
Refers to: specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:2187 in 8acc868. [](commit_id = 8acc8688c38f5db44352e7d4336e8e89927f0c87, deletion_comment = False)

majastrz commented:
>               }

This should always be returned. You can actually remove it from all response declarations.

---
Refers to: specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:2394 in 8acc868. [](commit_id = 8acc8688c38f5db44352e7d4336e8e89927f0c87, deletion_comment = False)

majastrz commented:
>               "$ref": "#/definitions/AssessmentOptions"

The GET on this path makes it look like `assessmentOptions` is a nested resource under `assessmentProjects`. However, the response doesn't match a typical collection GET. You have the following options here:

1. Use POST and name it `getAssessmentOptions`. It has the downside that Readers won't be able to get it unless you specify an authorizationActionMapping in the RP manifest.
2. Leave path as-is and change response to look like { "value": <array of objects>" }
3. Implement singleton pattern with path `/subscriptions/{subscriptionId}/resourcegroups/{resourceGroupName}/providers/Microsoft.Migrate/assessmentProjects/{projectName}/assessmentOptions/default`. You would need routes for single GET and collection GET here, but it is the cleanest from API consistency perspective.

---
Refers to: specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:2446 in 8acc868. [](commit_id = 8acc8688c38f5db44352e7d4336e8e89927f0c87, deletion_comment = False)

majastrz commented:
>     "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Migrate/assessmentProjects/{projectName}/groups/{groupName}/assessments/{assessmentName}/downloadUrl": {

Should this be called `getDownloadUrl`. Right now the name suggests that the action itself downloads a URL.

---
Refers to: specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:3134 in 8acc868. [](commit_id = 8acc8688c38f5db44352e7d4336e8e89927f0c87, deletion_comment = False)

majastrz commented on specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/examples/AssessedMachines_Get.json:
```diff
@@ -0,0 +1,82 @@
+          "monthlyStandardSSDStorageCost": 0.0,
+          "confidenceRatingInPercentage": 0.0,
+          "disks": {
+            "6000C299-02f5-d137-9bab-8a8ee7b192a0": {
```
>6000C299-02f5-d137-9bab-8a8ee7b192a0 [](start = 13, length = 36)

GUIDs typically have horrible usability. Is there another identifier we could use for disks? Does it map to something that you can find on the machine?

shsingla replied:
This is actually a UUID (Universally unique identifier) that we are getting from another service that is the discovery service. The discovery service gets it from VMware APIs so we are not adding anything from our end here. Moreover, this was the field that we found to be unique amongst all the properties, therefore, choosing this.

majastrz replied:
Ok.

---
In reply to: [325075829](https://github.com/Azure/azure-rest-api-specs/pull/7063#discussion_r325075829) [](ancestors = 325075829)

majastrz commented on specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/examples/AssessedMachines_Get.json:
```diff
@@ -0,0 +1,82 @@
+            }
+          },
+          "networkAdapters": {
+            "4000": {
```
>4000 [](start = 13, length = 4)

What does this refer to? How do you find it on the machine?

shsingla replied:
This is the nicId of the network adapter. This data we are getting from another service that is discovery service.

majastrz replied:
Makes sense.

---
In reply to: [325076337](https://github.com/Azure/azure-rest-api-specs/pull/7063#discussion_r325076337) [](ancestors = 325076337)

majastrz commented:
>         "properties": {

I see now that this is a different resource type. You should still consider adding x-ms-flatten back to make the SDK a cleaner.

---
In reply to: [532015874](https://github.com/Azure/azure-rest-api-specs/pull/7063#issuecomment-532015874) [](ancestors = 532015874)

---
Refers to: specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:162 in 8acc868. [](commit_id = 8acc8688c38f5db44352e7d4336e8e89927f0c87, deletion_comment = False)

majastrz commented:
>     "/subscriptions/{subscriptionId}/resourcegroups/{resourceGroupName}/providers/Microsoft.Migrate/assessmentProjects": {

I saw the response elsewhere. Makes sense.

---
In reply to: [532018223](https://github.com/Azure/azure-rest-api-specs/pull/7063#issuecomment-532018223) [](ancestors = 532018223,528015267)

---
Refers to: specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:2065 in f533bc5. [](commit_id = f533bc571046c13e488c2e4a8df0391a1ed6d3ef, deletion_comment = False)

majastrz reviewed (commented):
I added some comments and questions. Please take a look.

shsingla commented:
> downloadUrl

This is actually a post-call to get the download URL. So calling it get download URL might be confusing. Moreover the same was being used in our earlier API versions so we dont want to change it now as customers are already acquainted with this.

shsingla commented:
> > ```
> >           "$ref": "#/definitions/AssessmentOptions"
> > ```
> 
> The GET on this path makes it look like `assessmentOptions` is a nested resource under `assessmentProjects`. However, the response doesn't match a typical collection GET. You have the following options here:
> 
> 1. Use POST and name it `getAssessmentOptions`. It has the downside that Readers won't be able to get it unless you specify an authorizationActionMapping in the RP manifest.
> 2. Leave path as-is and change response to look like { "value": " }
> 3. Implement singleton pattern with path `/subscriptions/{subscriptionId}/resourcegroups/{resourceGroupName}/providers/Microsoft.Migrate/assessmentProjects/{projectName}/assessmentOptions/default`. You would need routes for single GET and collection GET here, but it is the cleanest from API consistency perspective.
> 
> Refers to: specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:2446 in [8acc868](https://github.com/Azure/azure-rest-api-specs/commit/8acc8688c38f5db44352e7d4336e8e89927f0c87). [](commit_id = [8acc868](https://github.com/Azure/azure-rest-api-specs/commit/8acc8688c38f5db44352e7d4336e8e89927f0c87), deletion_comment = False)

Actually this is not a collection/ list get and is only a get which returns object of type AssessmentOptions. Should we name this something else to avaoid the confusion?

shsingla commented:
> > ```
> >     "areAssessmentsRunning": {
> > ```
> 
> It's not required, but have you considered making it an enum? If it was named `assessmentStatus`, you could easily add more values in the future.
> 
> Refers to: specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:326 in [8acc868](https://github.com/Azure/azure-rest-api-specs/commit/8acc8688c38f5db44352e7d4336e8e89927f0c87). [](commit_id = [8acc868](https://github.com/Azure/azure-rest-api-specs/commit/8acc8688c38f5db44352e7d4336e8e89927f0c87), deletion_comment = False)

Actually we thought about it but we realized that this property might not have more than 2 values therefore we kept it as boolean.

nschonni commented on specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:
```diff
@@ -0,0 +1,3777 @@
+        },
+        "updatedTimestamp": {
+          "type": "string",
+          "decription": "Time when this collector was updated. Date-Time represented in ISO-8601 format.",
```
```suggestion
          "description": "Time when this collector was updated. Date-Time represented in ISO-8601 format.",
```

shsingla replied:
Done.

shsingla commented:
> > ```
> >     "properties": {
> > ```
> 
> x-ms-flatten was removed from this definition as well.
> 
> Refers to: specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:284 in [8acc868](https://github.com/Azure/azure-rest-api-specs/commit/8acc8688c38f5db44352e7d4336e8e89927f0c87). [](commit_id = [8acc868](https://github.com/Azure/azure-rest-api-specs/commit/8acc8688c38f5db44352e7d4336e8e89927f0c87), deletion_comment = False)

These are a completely new set of resources so the earlier swagger won't break. We can add x-ms-flatten but that will change our swagger SDK. This SDK is currently used by one of our internal teams and they will need to rewrite a good amount of code, so we will like to avoid it if not extremely necessary

shsingla commented:
> > ```
> >     "properties": {
> > ```
> 
> I see now that this is a different resource type. You should still consider adding x-ms-flatten back to make the SDK a cleaner.
> 
> In reply to: [532015874](https://github.com/Azure/azure-rest-api-specs/pull/7063#issuecomment-532015874) [](ancestors = 532015874)
> 
> Refers to: specification/migrate/resource-manager/Microsoft.Migrate/stable/2019-05-01/migrate.json:162 in [8acc868](https://github.com/Azure/azure-rest-api-specs/commit/8acc8688c38f5db44352e7d4336e8e89927f0c87). [](commit_id = [8acc868](https://github.com/Azure/azure-rest-api-specs/commit/8acc8688c38f5db44352e7d4336e8e89927f0c87), deletion_comment = False)

Not adding this for now as explained in a different comment.

shsingla commented:
> I added some comments and questions. Please take a look.

Took the comments please check.

shsingla committed:
Taking comments.

Taking comments.

shsingla commented:
> There were some existing comments that weren't resolved above as well

Took all comments.

majastrz commented:
> s-flatten but that will change our swagger SDK. This SDK is currently used by one of our internal teams an

You will need to suppress the violation you'll get from the PR checks on this, but it should be ok.

majastrz commented:
It's not about naming alone. It is more about where the segment is in the URL and which HTTP verb is being used with it. The `assessmentOptions` segment is in the place of a collections are as per the Resource Provider Contract, so a GET on it needs to return a collection response. You can either make it a collection or make it an action using options 1-3 I mentioned earlier.

---
In reply to: [532064744](https://github.com/Azure/azure-rest-api-specs/pull/7063#issuecomment-532064744) [](ancestors = 532064744)

majastrz commented:
Fair enough.

---
In reply to: [532064156](https://github.com/Azure/azure-rest-api-specs/pull/7063#issuecomment-532064156) [](ancestors = 532064156)

majastrz commented:
Got it.

---
In reply to: [528277311](https://github.com/Azure/azure-rest-api-specs/pull/7063#issuecomment-528277311) [](ancestors = 528277311)

yungezz commented:
hi @shsingla are  you still on this PR? could you pls give a update?

shsingla commented:
Yes @yungezz I am on this PR. Making some code changes to related to comments given here. Need to start the next iteration soon.

shsingla committed:
Making changes related to assessment options API and bumping up Api version to 2019-10-01.

1. AssessmentOptions list and Assessment options get singleton (with default name) is supported now.
2. Bumping up the API version. This was required because we dont want to break existing people using assessment options API.

shsingla committed:
Fixing CLI errors.

majastrz reviewed (approved):
Signed off from ARM side.

MyronFanQiu commented:
@shsingla Hello. There are two CI failures. Could you please fix them?

shsingla committed:
Fixing spell checks.

Fixing spell checks.

raych1 commented:
@shsingla , regard to model validation errors, one of the issue is example file `Assessments_Create.json` is not valid JSON which misses one closing brace for `parameters` section. Please fix that and re-run model check then you will see other detail errors.  Can you file an issue [here](https://github.com/Azure/oav/issues) for not exposing detail errors when example file is malformed.

shsingla committed:
Fixing model validations.

Fixing model validations.

shsingla committed:
Fixing CLI errors.

Fixing CLI errors.

MyronFanQiu commented:
@shsingla Hello. Did you solve the CI errors or need some helps?

MyronFanQiu commented:
@shsingla Hello. Did you solve the CI errors or need some helps?

MyronFanQiu commented:
@shsingla Hi. Any update?

MyronFanQiu commented:
@shsingla Hello! If this PR is not used anymore, I will close it in a few days.

MyronFanQiu commented:
@shsingla Hello. I close this PR for now. Feel free to re-open it.

shsingla committed:
Fixing some CLI issues.

Fixing some CLI issues.

MyronFanQiu commented:
@pilor Hello. Could you approve this PR?
